[DataPipe] implementing __len__ for fork (no valid length for demux)

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64215
* #63650
* #63649

Differential Revision: [D30648672](https://our.internmc.facebook.com/intern/diff/D30648672)

cc @VitalyFedyunin @ejguan

NivekT committed:
implementing __len__ for fork (no valid length for demux)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30648672).

ejguan reviewed (approved):
LGTM, can you please also add a test for `demux` that an Error will be raised.

NivekT committed:
Update on "[DataPipe] implementing __len__ for fork (no valid length for demux)"


Differential Revision: [D30648672](https://our.internmc.facebook.com/intern/diff/D30648672)

cc @VitalyFedyunin @ejguan

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30648672).

ejguan commented on test/test_datapipe.py:
```diff
@@ -776,6 +782,13 @@ def test_demux_datapipe(self):
 
+        # Test Case: __len__ not implemented
+        dp1, dp2 = input_dp.demux(num_instances=2, classifier_fn=lambda x: x % 2)
+        with self.assertRaises(TypeError):
```
IIRC, this is going to be `NotImplementedError`

NivekT replied:
I think Python throws a `TypeError` when `len()` is not implemented.

```
test/test_datapipe.py:792 (TestFunctionalIterDataPipe.test_error_some)
Traceback (most recent call last):
  File "/Users/ktse/pytorch/test/test_datapipe.py", line 802, in test_error_some
    len(dp1)
  File "/Users/ktse/pytorch/torch/utils/data/datapipes/iter/combining.py", line 162, in __len__
    return len(self.main_datapipe)
TypeError: object of type '_DemultiplexerIterDataPipe' has no len()

```

NivekT committed:
Update on "[DataPipe] implementing __len__ for fork (no valid length for demux)"


Differential Revision: [D30648672](https://our.internmc.facebook.com/intern/diff/D30648672)

cc @VitalyFedyunin @ejguan

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30648672).
